example_scripts/login_system-cache.pwn

```diff
@@ -46,224 +39,286 @@ new Player[MAX_PLAYERS][E_PLAYERS];
 new g_MysqlRaceCheck[MAX_PLAYERS];
 
 //dialog data
-enum 
+enum
 {
-	DIALOG_INVALID,
 	DIALOG_UNUSED,
-	
+
 	DIALOG_LOGIN,
-	DIALOG_REGISTER,
+	DIALOG_REGISTER
 };
 
-
-/*
- * SA-MP callbacks
- */
+main() {}
 
 public OnGameModeInit()
 {
-	mysql_log(LOG_ERROR | LOG_WARNING, LOG_TYPE_HTML); //logs errors and warnings into a nice .html log file
-	g_SQL = mysql_connect(SQL_HOST, SQL_USER, SQL_DB, SQL_PASS);
+	new MySQLOpt: option_id = mysql_init_options();
+
+	mysql_set_option(option_id, AUTO_RECONNECT, true); // It automatically reconnects when loosing connection to mysql server
+	mysql_set_option(option_id, MULTI_STATEMENTS, true); // Allows to execute multiple queries at once
+
+	g_SQL = mysql_connect(MYSQL_HOST, MYSQL_USER, MYSQL_PASSWORD, MYSQL_DATABASE, option_id); // AUTO_RECONNECT and MULTI_STATEMENTS are enabled for this connection handle only
+	if (g_SQL == MYSQL_INVALID_HANDLE || mysql_errno(g_SQL) != 0)
+	{
+		print("MySQL connection failed. Server is shutting down.");
+		SendRconCommand("exit"); // Close the server if there is no connection
+		return 1;
+	}
 	
+	print("MySQL connection is successful.");
 	SetupPlayerTable();
+	// If the table has been created, the "SetupPlayerTable" function does not have any purpose so you may remove it completely
 	return 1;
 }
 
 public OnGameModeExit()
 {
-	//save all player data before closing connection
-	for(new p=0; p < MAX_PLAYERS; ++p)
-		if(IsPlayerConnected(p))
-		    UpdatePlayerData(p);
-
-	mysql_close();
+	// save all player data before closing connection
+	for (new i = 0, j = GetPlayerPoolSize(); i <= j; i++) // GetPlayerPoolSize function was added in 0.3.7 version and gets the highest playerid currently in use on the server
+	{
+		if (IsPlayerConnected(i))
+		{
+			UpdatePlayerData(i);
+		}
+	}
+    
+	mysql_close(g_SQL);
 	return 1;
 }
 
 public OnPlayerConnect(playerid)
 {
 	g_MysqlRaceCheck[playerid]++;
-	//reset player data
-	for(new E_PLAYERS:e; e < E_PLAYERS; ++e)
+	
+	// reset player data
+	for (new E_PLAYERS: e; e < E_PLAYERS; ++e)
+	{
 	    Player[playerid][e] = 0;
+	}
 
-	new query[128];
 	GetPlayerName(playerid, Player[playerid][Name], MAX_PLAYER_NAME);
 
-	//send a query to recieve all the stored player data from the table
-	mysql_format(g_SQL, query, sizeof(query), "SELECT * FROM `players` WHERE `username` = '%e' LIMIT 1", Player[playerid][Name]);
+	// send a query to recieve all the stored player data from the table
+	new query[80];
```

Comment:
The query string should be bigger, since escaping a string may result in a longer output string (`2 * <source string length> + 1` worst case);